File: app/src/main/java/love/miao/yun/UpdateChecker.java
package love.miao.yun;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

/**
 * 从 GitHub Releases 检查是否有新版本。
 * 优先使用 GitHub API；若遇 403 限流则回退到页面解析。
 */
public class UpdateChecker {

    private static final String REPO = "Youzix-Star/NekoPlus";
    private static final String API_URL =
            "https://api.github.com/repos/" + REPO + "/releases/latest";
    private static final String RELEASES_PAGE =
            "https://github.com/" + REPO + "/releases/latest";
    private static final Handler MAIN = new Handler(Looper.getMainLooper());

    public interface Callback {
        void onUpdateAvailable(String latestVersion, String body, String apkUrl);
        void onNoUpdate();
        void onError(String message);
    }

    /** 异步检查更新，回调在主线程。 */
    public static void checkForUpdate(Context context, Callback callback) {
        String currentVersion = getCurrentVersion(context);
        if (currentVersion == null) {
            MAIN.post(() -> onError(callback, "无法获取当前版本"));
            return;
        }

        new Thread(() -> {
            try {
                URL url = new URL(API_URL);
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                try {
                    conn.setRequestMethod("GET");
                    conn.setConnectTimeout(10000);
                    conn.setReadTimeout(10000);
                    conn.setRequestProperty("Accept", "application/vnd.github.v3+json");

                    int code = conn.getResponseCode();
                    InputStream is = code >= 400 ? conn.getErrorStream() : conn.getInputStream();
                    String resp = readAll(is);
                    if (code >= 400) {
                        // API 限流或失败，回退到页面解析
                        checkViaPage(callback, currentVersion);
                        return;
                    }

                    JSONObject release = new JSONObject(resp);
                    String tag = release.optString("tag_name", "");
                    String body = release.optString("body", "");
                    // 去掉 tag 前缀 v
                    String latestVersion = tag.startsWith("v") ? tag.substring(1) : tag;

                    // 查找 APK 下载链接
                    String apkUrl = null;
                    JSONArray assets = release.optJSONArray("assets");
                    if (assets != null) {
                        for (int i = 0; i < assets.length(); i++) {
                            JSONObject asset = assets.getJSONObject(i);
                            String name = asset.optString("name", "");
                            if (name.endsWith(".apk") && !name.contains("debug")) {
                                apkUrl = asset.optString("browser_download_url", null);
                                break;
                            }
                        }
                    }

                    if (isNewer(latestVersion, currentVersion)) {
                        final String v = latestVersion;
                        final String b = body;
                        final String u = apkUrl;
                        MAIN.post(() -> callback.onUpdateAvailable(v, b, u));
                    } else {
                        MAIN.post(() -> callback.onNoUpdate());
                    }
                } finally {
                    conn.disconnect();
                }
            } catch (Exception e) {
                String msg = e.getMessage() == null ? e.toString() : e.getMessage();
                onError(callback, msg);
            }
        }).start();
    }

    /** 打开 GitHub Releases 页面供用户手动下载。 */
    public static void openReleasePage(Context context) {
        Intent intent = new Intent(Intent.ACTION_VIEW,
                Uri.parse("https://github.com/" + REPO + "/releases/latest"));
        context.startActivity(intent);
    }

    /**
     * 回退方案：访问 GitHub Releases 页面，从重定向 URL 中提取 tag_name。
     * 该页面会 302 重定向到 /releases/tag/vX.Y.Z，无需 API 认证，无速率限制。
     */
    private static void checkViaPage(Callback callback, String currentVersion) {
        new Thread(() -> {
            try {
                URL url = new URL(RELEASES_PAGE);
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                conn.setInstanceFollowRedirects(false); // 手动处理重定向
                try {
                    conn.setRequestMethod("GET");
                    conn.setConnectTimeout(10000);
                    conn.setReadTimeout(10000);

                    int code = conn.getResponseCode();
                    // 302 重定向到 /releases/tag/vX.Y.Z
                    if (code >= 300 && code < 400) {
                        String location = conn.getHeaderField("Location");
                        if (location != null && location.contains("/tag/")) {
                            String tag = location.substring(location.lastIndexOf("/") + 1);
                            String latestVersion = tag.startsWith("v") ? tag.substring(1) : tag;
                            if (isNewer(latestVersion, currentVersion)) {
                                final String v = latestVersion;
                                MAIN.post(() -> callback.onUpdateAvailable(v, "", null));
                            } else {
                                MAIN.post(() -> callback.onNoUpdate());
                            }
                            return;
                        }
                    }
                    // 无法解析，报错
                    onError(callback, "无法获取版本信息");
                } finally {
                    conn.disconnect();
                }
            } catch (Exception e) {
                String msg = e.getMessage() == null ? e.toString() : e.getMessage();
                onError(callback, msg);
            }
        }).start();
    }

    /** 从 PackageManager 获取当前版本名（如 "1.7.8"）。 */
    private static String getCurrentVersion(Context context) {
        try {
            return context.getPackageManager()
                    .getPackageInfo(context.getPackageName(), 0).versionName;
        } catch (PackageManager.NameNotFoundException e) {
            return null;
        }
    }

    /** 比较两个语义化版本号，返回 latest > current。 */
    private static boolean isNewer(String latest, String current) {
        String[] la = latest.split("\\.");
        String[] cu = current.split("\\.");
        int len = Math.max(la.length, cu.length);
        for (int i = 0; i < len; i++) {
            int l = i < la.length ? parsePart(la[i]) : 0;
            int c = i < cu.length ? parsePart(cu[i]) : 0;
            if (l > c) return true;
            if (l < c) return false;
        }
        return false;
    }

    private static int parsePart(String s) {
        try {
            return Integer.parseInt(s.replaceAll("[^0-9]", ""));
        } catch (Exception e) {
            return 0;
        }
    }

    private static String readAll(InputStream is) throws Exception {
        if (is == null) return "";
        BufferedReader reader = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            sb.append(line).append("\n");
        }
        reader.close();
        return sb.toString();
    }

    private static void onError(Callback callback, String msg) {
        MAIN.post(() -> callback.onError(msg));
    }
}

File: app/src/main/java/love/miao/yun/util/UpdateChecker.kt
/*
 * Copyright 2026, Youzix-Star
 * SPDX-License-Identifier: GPL-3.0-only
 *
 * Ported from NekoNeko (top.youzix.nekoneko), UpdateChecker.java.
 */

package love.miao.yun.util

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import love.miao.yun.BuildConfig
import org.json.JSONArray
import org.json.JSONObject

/** What a version check found. */
sealed interface UpdateResult {
    /** A newer release exists. [apkUrl] is null when the release carries no APK asset. */
    data class Available(
        val version: String,
        val notes: String,
        val apkUrl: String?,
    ) : UpdateResult

    data object UpToDate : UpdateResult

    data class Failed(val message: String) : UpdateResult
}

/**
 * Checks GitHub Releases for a newer build.
 *
 * The API is tried first; when it answers with an error — rate limiting, most often — the check
 * falls back to the releases page and reads the tag out of its redirect, which needs no token and
 * has no quota. That fallback is the difference between "check for updates" working and quietly
 * failing for the one user who taps it twice.
 *
 * A pre-release build asks for every release and a stable build only for the latest stable one:
 * `/releases/latest` deliberately ignores pre-releases, so an alpha that consulted it would never
 * see the next alpha — and a stable build that consulted the full list would be offered an alpha.
 */
object UpdateChecker {
    private const val REPO = "Youzix-Star/NekoPlus"

    /** The newest stable release; GitHub deliberately leaves pre-releases out of this one. */
    private const val API_LATEST = "https://api.github.com/repos/$REPO/releases/latest"

    /** Every release, newest first — the only way a pre-release can see other pre-releases. */
    private const val API_LIST = "https://api.github.com/repos/$REPO/releases?per_page=20"

    private const val RELEASES_PAGE = "https://github.com/$REPO/releases/latest"

    private val main = Handler(Looper.getMainLooper())

    /** Runs the check off the main thread and calls back on it. */
    fun check(onResult: (UpdateResult) -> Unit) {
        val current = BuildConfig.VERSION_NAME
        Thread {
            val result = try {
                viaApi(current)
            } catch (error: Exception) {
                // A failed request is not the end of the story: try the page fallback first.
                runCatching { viaPage(current) }.getOrElse {
                    UpdateResult.Failed(error.message ?: error.toString())
                }
            }
            main.post { onResult(result) }
        }.apply { isDaemon = true }.start()
    }

    /** Opens the release page, for when the user would rather grab the APK by hand. */
    fun openReleasePage(context: Context) {
        runCatching {
            context.startActivity(
                Intent(Intent.ACTION_VIEW, Uri.parse(RELEASES_PAGE))
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK),
            )
        }
    }

    private fun viaApi(current: String): UpdateResult {
        // Which releases this build is allowed to see follows from what it is: a pre-release looks
        // at every release, because that is the channel it is on, while a stable build only ever
        // sees stable ones — otherwise an alpha would be pushed at everyone.
        val trackPreReleases = current.contains('-')
        val releases = fetchReleases(if (trackPreReleases) API_LIST else API_LATEST, current)
            ?: return viaPage(current)
        val newest = releases
            .filter { release ->
                !release.optBoolean("draft") &&
                    (trackPreReleases || !release.optBoolean("prerelease"))
            }
            .maxByOrNull { release -> release.optString("tag_name", "").removePrefix("v") }
            ?: return UpdateResult.UpToDate

        val version = newest.optString("tag_name", "").removePrefix("v")
        if (!isNewer(version, current)) return UpdateResult.UpToDate

        var apkUrl: String? = null
        newest.optJSONArray("assets")?.let { assets ->
            for (index in 0 until assets.length()) {
                val asset = assets.optJSONObject(index) ?: continue
                val name = asset.optString("name", "")
                if (name.endsWith(".apk") && !name.contains("debug")) {
                    apkUrl = asset.optString("browser_download_url", null)
                    break
                }
            }
        }
        return UpdateResult.Available(
            version = version,
            notes = newest.optString("body", ""),
            apkUrl = apkUrl,
        )
    }

    /** Both endpoints answer with JSON; one gives an object, the other an array. */
    private fun fetchReleases(url: String, current: String): List<JSONObject>? {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 10_000
            readTimeout = 10_000
            setRequestProperty("Accept", "application/vnd.github.v3+json")
        }
        return try {
            if (connection.responseCode >= 400) return null
            val body = readAll(connection.inputStream)
            if (body.trimStart().startsWith("[")) {
                val array = JSONArray(body)
                (0 until array.length()).mapNotNull { array.optJSONObject(it) }
            } else {
                listOf(JSONObject(body))
            }
        } finally {
            connection.disconnect()
        }
    }

    private fun viaPage(current: String): UpdateResult {
        val connection = (URL(RELEASES_PAGE).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 10_000
            readTimeout = 10_000
            instanceFollowRedirects = false
        }
        return try {
            val code = connection.responseCode
            val location = if (code in 300..399) connection.getHeaderField("Location") else null
            if (location != null && location.contains("/tag/")) {
                val version = location.substringAfterLast("/").removePrefix("v")
                if (isNewer(version, current)) {
                    UpdateResult.Available(version = version, notes = "", apkUrl = null)
                } else {
                    UpdateResult.UpToDate
                }
            } else if (current.contains('-')) {
                // This fallback reads /releases/latest, which by definition skips pre-releases,
                // so for a pre-release build the honest answer is "cannot tell", not "up to date".
                UpdateResult.Failed("接口受限，暂时无法确认预发布版本")
            } else {
                UpdateResult.Failed("无法获取版本信息")
            }
        } finally {
            connection.disconnect()
        }
    }

    private fun readAll(input: InputStream?): String {
        if (input == null) return ""
        return BufferedReader(InputStreamReader(input, StandardCharsets.UTF_8)).use { reader ->
            reader.readText()
        }
    }
}

/**
 * Whether [latest] is a later release than [current].
 *
 * Versions are compared on their numeric core first, and only then on the pre-release suffix —
 * so `2.0.0` counts as newer than `2.0.0-alpha.2`, which the plain "strip every non-digit"
 * comparison the original used would have missed, leaving alpha users stuck for good.
 */
internal fun isNewer(latest: String, current: String): Boolean {
    val (latestCore, latestPre) = splitVersion(latest)
    val (currentCore, currentPre) = splitVersion(current)

    for (index in 0 until maxOf(latestCore.size, currentCore.size)) {
        val l = latestCore.getOrElse(index) { 0 }
        val c = currentCore.getOrElse(index) { 0 }
        if (l != c) return l > c
    }

    // Same numbers: a finished release beats any pre-release of it.
    if (latestPre == null && currentPre != null) return true
    if (latestPre != null && currentPre == null) return false
    if (latestPre == null && currentPre == null) return false
    return comparePreRelease(latestPre!!, currentPre!!) > 0
}

/**
 * Puts the separators a version comparison needs into a version a person wrote.
 *
 * `2.0.2Beta1` is a perfectly normal thing to type, and without this it reads as the number 2.0.21
 * with no pre-release at all — which is how a beta build ends up being told there is a newer
 * version, or never told about the next one.
 */
private fun separateVersion(version: String): String {
    val out = StringBuilder(version.length + 4)
    version.forEachIndexed { index, character ->
        if (index > 0) {
            val previous = version[index - 1]
            val letterAfterDigit = previous.isDigit() && character.isLetter()
            val digitAfterLetter = previous.isLetter() && character.isDigit()
            if (letterAfterDigit) out.append('-') else if (digitAfterLetter) out.append('.')
        }
        out.append(character)
    }
    return out.toString()
}

private fun splitVersion(version: String): Pair<List<Int>, String?> {
    // "2.0.1 Beta 1" is a version name a person would write; the comparison wants a separator.
    val cleaned = separateVersion(version.trim().replace(' ', '-').removePrefix("v"))
    val core = cleaned.substringBefore('-')
    val pre = cleaned.substringAfter('-', missingDelimiterValue = "").ifBlank { null }
    val numbers = core.split('.').map { part ->
        part.filter { it.isDigit() }.toIntOrNull() ?: 0
    }
    return numbers to pre
}

/** `alpha.2` < `alpha.10` < `beta.1` < `rc.1`, with plain words ranking by name. */
private fun comparePreRelease(latest: String, current: String): Int {
    val left = latest.split('.')
    val right = current.split('.')
    for (index in 0 until maxOf(left.size, right.size)) {
        val l = left.getOrNull(index) ?: return -1
        val c = right.getOrNull(index) ?: return 1
        val ln = l.toIntOrNull()
        val cn = c.toIntOrNull()
        val comparison = when {
            ln != null && cn != null -> ln.compareTo(cn)
            ln != null -> -1
            cn != null -> 1
            else -> rank(l).compareTo(rank(c)).takeIf { it != 0 } ?: l.compareTo(c)
        }
        if (comparison != 0) return comparison
    }
    return 0
}

private fun rank(word: String): Int = when (word.lowercase()) {
    "alpha", "a" -> 0
    "beta", "b" -> 1
    "rc" -> 2
    else -> 3
}
